swarm/storage/mru/updaterequest.go

```diff
@@ -0,0 +1,277 @@
+	}
+
+	mr := &UpdateRequest{
+		SignedResourceUpdate: SignedResourceUpdate{
+			resourceUpdate: resourceUpdate{
+				version:   mj.Version,
+				period:    mj.Period,
+				multihash: mj.Multihash,
+			},
+		},
+		resourceMetadata: resourceMetadata{
+			name:      mj.Name,
+			frequency: mj.Frequency,
+			startTime: mj.StartTime,
+		},
+	}
+
+	ownerAddrBytes, err := hexutil.Decode(mj.OwnerAddr)
+	if err != nil || len(ownerAddrBytes) != common.AddressLength {
+		return nil, NewError(ErrInvalidValue, "Cannot decode ownerAddr")
+	}
+	copy(mr.ownerAddr[:], ownerAddrBytes)
+
+	if mj.Data != "" {
+		mr.data, err = hexutil.Decode(mj.Data)
+		if err != nil {
+			return nil, NewError(ErrInvalidValue, "Cannot decode data")
+		}
+	}
+	if mj.RootAddr != "" {
+		mr.rootAddr, err = hexutil.Decode(mj.RootAddr)
+		if err != nil {
+			return nil, NewError(ErrInvalidValue, "Cannot decode rootAddr")
+		}
+	}
+
+	if mj.MetaHash != "" {
+		mr.metaHash, err = hexutil.Decode(mj.MetaHash)
+		if err != nil {
+			return nil, NewError(ErrInvalidValue, "Cannot decode metaHash")
+		}
+	}
+
+	if mj.Signature != "" {
+		sigBytes, err := hexutil.Decode(mj.Signature)
+		if err != nil || len(sigBytes) != signatureLength {
+			return nil, NewError(ErrInvalidSignature, "Cannot decode signature")
+		}
+		mr.signature = new(Signature)
+		copy(mr.signature[:], sigBytes)
+	}
+	return mr, nil
+}
+
+func NewUpdateRequest(name string, frequency, startTime uint64, ownerAddr common.Address, data []byte, multihash bool) (*UpdateRequest, error) {
+	if !isSafeName(name) {
+		return nil, NewError(ErrInvalidValue, fmt.Sprintf("Invalid name: '%s' when creating NewMruRequest", name))
+	}
+
+	if startTime == 0 {
+		startTime = uint64(time.Now().Unix())
+	}
+
+	updateRequest := &UpdateRequest{
+		SignedResourceUpdate: SignedResourceUpdate{
+			resourceUpdate: resourceUpdate{
+				version:   1,
+				period:    1,
+				data:      data,
+				multihash: multihash,
+			},
+		},
+		resourceMetadata: resourceMetadata{
+			name:      name,
+			frequency: frequency,
+			startTime: startTime,
+			ownerAddr: ownerAddr,
+		},
+	}
+
+	updateRequest.rootAddr, updateRequest.metaHash, _ = updateRequest.resourceMetadata.hash()
+
+	return updateRequest, nil
+}
+
+func (mr *UpdateRequest) Verify() error {
+	key := resourceHash(mr.period, mr.version, mr.rootAddr)
+	digest := keyDataHash(key, mr.metaHash, mr.data)
+	// get the address of the signer (which also checks that it's a valid signature)
+	addr, err := getAddressFromDataSig(digest, *mr.signature)
+	if err != nil {
+		return err
+	}
+	if addr != mr.ownerAddr {
+		return NewError(ErrInvalidSignature, "Signature address does not match with ownerAddr")
+	}
+	_, err = mr.SignedResourceUpdate.Verify(key)
+	return err
+}
+
+func (mr *UpdateRequest) Sign(signer Signer) error {
+	if err := mr.SignedResourceUpdate.Sign(signer); err != nil {
+		return err
+	}
+	mr.ownerAddr = signer.Address()
+	return nil
+}
+
+func (mr *UpdateRequest) Frequency() uint64 {
```

Comment:
I don't really understand your logic here.

Generally I like to avoid adding function calls when it's not needed.

Reply:
Well, the frequency is read-only really. When you create a resource it is set in stone.

We are exposing this structure to the end user, so my point of view is that it has to be clear what can and what can't be modified.

Here are the two ways an `UpdateRequest` is used:

1.- To create a resource:

To create a resource, you invoke `mru.NewUpdateRequest` (later renamed to `mru.NewCreateRequest`), passing the following parameters:
`(name string, frequency, startTime uint64, ownerAddr common.Address, data []byte, multihash bool)`

This returns a structure that you can `.Sign()` and post to the `Handler`. The object contains everything needed to create + do the first update of a resource.

2.- To Update a resource:

Call `Handler.NewUpdateRequest`. This will return an `UpdateRequest` that is prefilled with everything except the new data. So you:

Call `Handler.NewUpdateRequest()` to obtain a template (will be prefilled with the right period, version, etc)
Call `UpdateRequest.SetData()` to actually put the data you want to send off
Call `UpdateRequest.Sign()`
Call `Handler.Update()` to submit the update back to the handler.

That is, once the resource is created you really can't change things such as `rootAddr` or `frequency`, so you shouldn't really think you can, because submitting an `UpdateRequest` to a `Handler.Update` with an altered frequency won't really change the frequency: that field will be ignored.

In terms of efficiency, that function call will be inlined by the compiler, i.e., it won't really be a `CALL`, will be compiled to a `MOV` directly.

In any case, I'll be refactoring this a bit due to our previous conversation regarding having a CLI method that creates and another that updates, separately. I think this separation should also be at this level.